Add examples of build operations

hongkailiu opened the pull request.

mffiedler commented on examples/build_operation/README.md:
```diff
@@ -0,0 +1,31 @@
+# Build operations
+
+This example program demonstrates the fundamental operations for managing on
```
for managing Build resources

mffiedler commented on examples/build_operation/main.go:
```diff
@@ -0,0 +1,96 @@
+	"os"
+	"fmt"
+)
+
```
Suggest adding a short comment before each section //List builds,   //Retrieve build,   //Start build   etc

hongkailiu replied:
Done.

mffiedler commented on examples/build_operation/main.go:
```diff
@@ -0,0 +1,96 @@
+	} else {
+		fmt.Printf("Found build %s in namespace %s\n", build, namespace)
+		fmt.Printf("Found build %s in status %+v\n", build, myBuild.Status)
+	}
```
In addition to the raw printout of the build, can we show how to extract select fields and print them out nicely?  Maybe build name, start time, duration and status/phgase

hongkailiu replied:
Done

hongkailiu commented on INSTALL.md:
```diff
@@ -0,0 +1,31 @@
+# Installing client-go
```
Add a paragraph to clarify this.

hongkailiu commented on examples/build_operation/README.md:
```diff
@@ -0,0 +1,31 @@
+# Build operations
```
Done

mffiedler commented:
@hongkailiu  I think you're ready to squash and get some additional reviews/comments.

hongkailiu commented:
@deads2k This PR is ready for review. Thanks.

deads2k commented on INSTALL.md:
````diff
@@ -0,0 +1,34 @@
+vendor/. Then run the following command in the same directory as `glide.yaml`:
+
+```sh
+$ glide up -v
````
recommend `--strip-vendor` too.

hongkailiu replied:
Done

deads2k commented on examples/build/README.md:
````diff
@@ -0,0 +1,35 @@
+
+```sh
+### Load dependencies if not done yet (required only once)
+$ glide up -v
````
This surprises me.  I was fairly sure we include everything in our vendor tree already.

hongkailiu replied:
If we do not do the command, we would get this when building:

[hongkliu@hongkliu client-go]$ go build -o examples/build/app ./examples/build/
vendor/k8s.io/client-go/tools/clientcmd/auth_loaders.go:26:2: cannot find package "github.com/howeyc/gopass" in any of:
        /home/hongkliu/repo/go/src/github.com/openshift/client-go/vendor/github.com/howeyc/gopass (vendor tree)
        /home/hongkliu/tool/go/src/github.com/howeyc/gopass (from $GOROOT)
        /home/hongkliu/repo/go/src/github.com/howeyc/gopass (from $GOPATH)
vendor/k8s.io/client-go/tools/clientcmd/client_config.go:28:2: cannot find package "github.com/imdario/mergo" in any of:
        /home/hongkliu/repo/go/src/github.com/openshift/client-go/vendor/github.com/imdario/mergo (vendor tree)
        /home/hongkliu/tool/go/src/github.com/imdario/mergo (from $GOROOT)
        /home/hongkliu/repo/go/src/github.com/imdario/mergo (from $GOPATH)

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	buildv1 "github.com/openshift/client-go/build/clientset/versioned/typed/build/v1"
+	"path/filepath"
```
standard imports go first and then a newline, then kube, then a newline, then openshift

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	flag.Parse()
+
+	// use the current context in kubeconfig
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
```
This doesn't provide a way to register the flags?

hongkailiu replied:
I copied the code from k8s go-client example:
https://github.com/kubernetes/client-go/blob/master/examples/out-of-cluster-client-configuration/main.go
Also searched a bit on the code there, did not find any.

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	"fmt"
+)
+
+func main() {
```
Have your main call a function that returns an error.  It makes the code below copy/pasteable if you do that.

hongkailiu replied:
Done

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	namespace := "testproject"
+	// get all builds
+	builds, err := buildV1Client.Builds(namespace).List(metav1.ListOptions{})
+
```
delete this newline

hongkailiu replied:
Done

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	// get a specific build
+	build := "cakephp-ex-1"
+	myBuild, err := buildV1Client.Builds(namespace).Get(build, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
```
switch/case to match project style.

hongkailiu replied:
After adding start() function with error returned, I thought it could be simplified by just treating all types of error by panic.

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	myBuildConfig, err := buildV1Client.BuildConfigs(namespace).Get(buildConfig, metav1.GetOptions{})
+	if err != nil {
+		panic(err.Error())
+	} else {
```
no else needed

hongkailiu replied:
Done

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+
+		if err != nil {
+			panic(err.Error())
+		} else {
```
no else needed.

hongkailiu replied:
Done.

deads2k commented:
Add building the examples as a target to our makefile and make sure that travis executes it.

hongkailiu commented:
@deads2k 
The build of example on travis has similar error:
https://travis-ci.org/openshift/client-go

Should we add it into dep?

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,98 @@
+	build := "cakephp-ex-1"
+	myBuild, err := buildV1Client.Builds(namespace).Get(build, metav1.GetOptions{})
+	if err != nil {
+		panic(err.Error())
```
don't panic, return the errors

hongkailiu replied:
Sorry, missed 3 of them.
Done

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,98 @@
+	buildConfig := "cakephp-ex"
+	myBuildConfig, err := buildV1Client.BuildConfigs(namespace).Get(buildConfig, metav1.GetOptions{})
+	if err != nil {
+		panic(err.Error())
```
no panic in this function (throughout)

deads2k commented on examples/build/main.go:
```diff
@@ -0,0 +1,98 @@
+func main() {
+	err := start()
+	if err != nil {
+		panic(err.Error())
```
```
fmt.Fprintf(os.Stderr, "error: %v", err)
os.Exit(1)
```

hongkailiu replied:
Done!

deads2k commented:
What does `make update-deps` do?  Does it bring in the new dependencies?

hongkailiu commented:
Would this mean it does bring in the new dep? Sorry for the newbie question.

```sh
$ make update-deps
...
$ git diff glide.lock
diff --git a/glide.lock b/glide.lock
index 2e3f690..e1835ab 100644
--- a/glide.lock
+++ b/glide.lock
@@ -1,5 +1,5 @@
 hash: 6da94a04d17790083fe43232e1633957692729a49ee9d4a223dbc8225915b94c
-updated: 2017-12-15T12:18:26.934482511-05:00
+updated: 2018-01-25T17:01:04.417821963-05:00
 imports:
 - name: github.com/davecgh/go-spew
   version: 782f4967f2dc4564575ca782fe2d04090b5faca8
@@ -25,7 +25,7 @@ imports:
   - proto
   - sortkeys
 - name: github.com/golang/glog
-  version: 23def4e6c14b4da8ac2ed8007337bc5eb5007998
+  version: 44145f04b68cf362d9c4df2182967c2275eaefed
 - name: github.com/golang/protobuf
   version: 1643683e1b54a9e88ad26d98f81400c8c9d9f4f9
   subpackages:
@@ -52,6 +52,10 @@ imports:
   version: a0d98a5f288019575c6d1f4bb1573fef2d1fcdc4
   subpackages:
   - simplelru
+- name: github.com/howeyc/gopass
+  version: bf9dde6d0d2c004a008c27aaee91170c786f6db8
+- name: github.com/imdario/mergo
+  version: 6633656539c1639d9d78127b7d47c622b5d7b6dc
 - name: github.com/json-iterator/go
   version: 36b14963da70d11297d313183d7e6388c8510e1e
 - name: github.com/juju/ratelimit
@@ -63,7 +67,7 @@ imports:
   - jlexer
   - jwriter
 - name: github.com/openshift/api
-  version: 3922a2604003bc9fba95f09e17581e9917270a65
+  version: 0d921e363e951d89f583292c60d013c318df64dc
   subpackages:
   - apps/v1
   - authorization/v1
@@ -88,6 +92,10 @@ imports:
   version: 5bd2802263f21d8788851d5305584c82a5c75d7e
 - name: github.com/spf13/pflag
   version: 9ff6c6923cfffbcd502984b8e0c80539a94968b7
+- name: golang.org/x/crypto
+  version: 81e90905daefcd6fd217b62423c0908922eadb30
+  subpackages:
+  - ssh/terminal
 - name: golang.org/x/net
   version: 1c05540f6879653db88113bc4a2b70aec4bd491f
   subpackages:
@@ -97,6 +105,11 @@ imports:
   - http2/hpack
   - idna
   - lex/httplex
+- name: golang.org/x/sys
+  version: 95c6576299259db960f6c5b9b69ea52422860fce
+  subpackages:
+  - unix
+  - windows
 - name: golang.org/x/text
   version: b19bf474d317b857955b12035d2c5acb57ce8b01
   subpackages:
@@ -116,7 +129,7 @@ imports:
 - name: gopkg.in/yaml.v2
   version: 53feefa2559fb8dfa8d81baad31be332c97d6c77
 - name: k8s.io/api
-  version: af4bc157c3a209798fc897f6d4aaaaeb6c2e0d6a
+  version: 006a217681ae70cbacdd66a5e2fca1a61a8ff28e
   subpackages:
   - admissionregistration/v1alpha1
   - admissionregistration/v1beta1
@@ -147,7 +160,7 @@ imports:
   - storage/v1alpha1
   - storage/v1beta1
 - name: k8s.io/apimachinery
-  version: 180eddb345a5be3a157cea1c624700ad5bd27b8f
+  version: 68f9c3a1feb3140df59c67ced62d3a5df8e6c9c2
   subpackages:
   - pkg/api/errors
   - pkg/api/meta
@@ -188,7 +201,7 @@ imports:
   - pkg/watch
   - third_party/forked/golang/reflect
 - name: k8s.io/client-go
-  version: 78700dec6369ba22221b72770783300f143df150
+  version: 9389c055a838d4f208b699b3c7c51b70f2368861
   subpackages:
   - discovery
   - discovery/fake
@@ -197,17 +210,22 @@ imports:
   - rest
   - rest/watch
   - testing
+  - tools/auth
   - tools/cache
+  - tools/clientcmd
   - tools/clientcmd/api
+  - tools/clientcmd/api/latest
+  - tools/clientcmd/api/v1
   - tools/metrics
   - tools/pager
   - transport
   - util/buffer
   - util/cert
   - util/flowcontrol
+  - util/homedir
   - util/integer
 - name: k8s.io/code-generator
-  version: 0ea404e72920cc2978c0bfcc05a37528e01e5bcf
+  version: ff3e9afce6dc489cc817ade6556ccf62373dfd56
   repo: git@github.com:openshift/kubernetes-code-generator
 - name: k8s.io/gengo
   version: c2e273a26d3fc59b60cf6de2fb80eace7402de6e

```

deads2k commented:
> Would this mean it does bring in the new dep? Sorry for the newbie question.

Yes.  It means more repos are picked up by the vendoring tool.  I think its ok to gather these.

hongkailiu commented:
@deads2k 
Rebase: Done. Travis: Passed with updated deps.
The following line is removed (in examples/build/README.md) since the dependencies are updated:
$ glide up -v

wanghaoran1988 commented on examples/build/main.go:
```diff
@@ -0,0 +1,99 @@
+	fmt.Printf("There are %d builds in project %s\n", len(builds.Items), namespace)
+	// List names of all builds
+	for i, build := range builds.Items {
+		fmt.Sprintf("index %d: Name of the build: %s", i, build.Name)
```
s/Sprintf/Printf

hongkailiu replied:
Done.
Thanks.

deads2k commented:
lgtm, please squash for merge
